### Import thư viện

In [4]:
import pandas as pd
import numpy as np

---
### Khám phá dữ liệu

Load dữ liệu

In [20]:
df = pd.read_csv('chemicals-in-cosmetics-3.csv')
df.head()

index  CDPHId                                       ProductName  CSFId  \
0      0       2  ULTRA COLOR RICH EXTRA PLUMP LIPSTICK-ALL SHADES    NaN   
1      1       3                        Glover's Medicated Shampoo    NaN   
2      2       3                        Glover's Medicated Shampoo    NaN   
3      3       4          PRECISION GLIMMER EYE LINER-ALL SHADES �    NaN   
4      4       5       AVON BRILLIANT SHINE LIP GLOSS-ALL SHADES �    NaN   

   CSF  CompanyId          CompanyName BrandName  PrimaryCategoryId  \
0  NaN          4         New Avon LLC      AVON                 44   
1  NaN        338  J. Strickland & Co.  Glover's                 18   
2  NaN        338  J. Strickland & Co.  Glover's                 18   
3  NaN          4         New Avon LLC      AVON                 44   
4  NaN          4         New Avon LLC      AVON                 44   

                     PrimaryCategory  ...   CasNumber ChemicalId  \
0    Makeup Products (non-permanent)  ...  13463-67-7          6   
1  Hair Care Products (non-coloring)  ...  65996-92-1          4   
2  Hair Care Products (non-coloring)  ...    140-67-0          5   
3    Makeup Products (non-permanent)  ...  13463-67-7          7   
4    Makeup Products (non-permanent)  ...  13463-67-7          8   

             ChemicalName InitialDateReported  MostRecentDateReported  \
0        Titanium dioxide          06/17/2009              08/28/2013   
1  Distillates (coal tar)          07/01/2009              07/01/2009   
2               Estragole          07/01/2009              07/01/2009   
3        Titanium dioxide          07/09/2009              08/28/2013   
4        Titanium dioxide          07/09/2009              08/28/2013   

  DiscontinuedDate ChemicalCreatedAt ChemicalUpdatedAt ChemicalDateRemoved  \
0       02/01/2011        07/09/2009        07/09/2009                 NaN   
1              NaN        07/01/2009        07/01/2009                 NaN   
2              NaN        07/02/2009        07/02/2009                 NaN   
3              NaN        07/09/2009        07/09/2009                 NaN   
4       02/01/2011        07/09/2009        07/09/2009                 NaN   

  ChemicalCount  
0             1  
1             2  
2             2  
3             1  
4             1  

[5 rows x 23 columns]

Kiểm tra dữ liệu null

In [51]:
# tính số giá trị thiếu của cột s
def missing_num(s):
    return s.isnull().sum()

# Tính số tỉ lệ phần trăm giá trị thiếu của cột s 
def missing_ratio(s):
    return s.isnull().sum()/len(s)*100

# Tính số giá trị khác nhau (không tính giá trị thiếu) của cột s
def n_diff_vals(s):
    return s.dropna().unique().size

# Các giá trị khác nhau (không xét giá trị thiếu) của cột s
def diff_vals(s):
    return s.dropna().unique()

# Tên các cột thuộc kiểu dữ liệu dạng phân loại (category)
cate_index = ['ProductName', 'CSF', 'CompanyName', 'BrandName', 'PrimaryCategory',\
            'SubCategory', 'CasNumber', 'ChemicalName']

cate_columns = ['missing_num','missing_ratio', 'n_diff_vals', 'diff_vals']

cate_data = np.array([[
                missing_num(df[i]),
                missing_ratio(df[i]),
                n_diff_vals(df[i]),
                diff_vals(df[i]),
            ] for i in cate_index], dtype = object)

cate_col_profiles_df = pd.DataFrame(data=cate_data, columns=cate_columns, index=cate_index)

In [52]:
cate_col_profiles_df

missing_num missing_ratio n_diff_vals  \
ProductName               0           0.0       33638   
CSF                   34340      30.04427       34258   
CompanyName               0           0.0         606   
BrandName               216       0.18898        2711   
PrimaryCategory           0           0.0          13   
SubCategory               0           0.0          89   
CasNumber              6396      5.595898         125   
ChemicalName              0           0.0         123   

                                                         diff_vals  
ProductName      [ULTRA COLOR RICH EXTRA PLUMP LIPSTICK-ALL SHA...  
CSF              [5858-81-1, D&C RED 7 CALCIUM LAKE, D&C RED 28...  
CompanyName      [New Avon LLC, J. Strickland & Co., OPI PRODUC...  
BrandName        [AVON, Glover's, OPI, ABSOLUTE, ABSOLUTE FX, G...  
PrimaryCategory  [Makeup Products (non-permanent), Hair Care Pr...  
SubCategory      [Lip Color - Lipsticks, Liners, and Pencils, H...  
CasNumber        [13463-67-7, 65996-92-1, 140-67-0, 68603-42-9,...  
ChemicalName     [Titanium dioxide, Distillates (coal tar), Est...


<br>Số lượng giá trị thiếu (missing_num)
- Có 34340 dòng bị mất giá trị ở cột 'CSF'
- Có 216 dòng bị mất giá trị ở cột 'BrandName'
- Có 6396 dòng bị mất giá trị ở cột 'CasNumber'

<br>Tỉ lệ phần trăm giá trị thiếu (missing_ratio)
- Có 30% dòng bị mất giá trị ở cột 'CSF'
- Có 0.2% dòng bị mất giá trị ở cột 'BrandName'
- Có 5.6% dòng bị mất giá trị ở cột 'CasNumber'
